patroni/config_generator.py

```diff
@@ -0,0 +1,290 @@
+        if val:
+            su_params[conn_param] = val
+    patroni_env_su_username = ((config.get('authentication') or {}).get('superuser') or {}).get('username')
+    patroni_env_su_pwd = ((config.get('authentication') or {}).get('superuser') or {}).get('password')
+    # because we use "username" in the config for some reason
+    su_params['username'] = su_params.pop('user', patroni_env_su_username) or getuser()
+    su_params['password'] = su_params.get('password', patroni_env_su_pwd) or getpass('Please enter the user password:')
+
+    from . import psycopg
+    try:
+        conn = psycopg.connect(dsn=dsn, password=su_params['password'])
+    except psycopg.Error as e:
+        sys.exit(f'Failed to establish PostgreSQL connection: {e}')
+
+    with conn.cursor() as cur:
+        cur.execute("SELECT 1 FROM pg_roles WHERE rolname=%s AND rolsuper", (su_params['username'],))
+        if cur.rowcount < 1:
+            sys.exit('The provided user does not have superuser privilege')
+
+        required_params = ['hba_file', 'ident_file', 'config_file',
+                           'data_directory'] + list(ConfigHandler.CMDLINE_OPTIONS.keys())
+        cur.execute("SELECT name, current_setting(name) FROM pg_settings "
+                    "WHERE context <> 'internal' "
+                    "AND source IN ('configuration file', 'command line', 'environment variable') "
+                    "AND category <> 'Write-Ahead Log / Recovery Target' "
+                    "AND setting <> '(disabled)' "
+                    "OR name = ANY(%s)", (required_params,))
+
+        helper_dict = dict.fromkeys(['port', 'listen_addresses'])
+        # adjust values
+        for p, v in cur.fetchall():
+            if p == 'data_directory':
+                config['postgresql']['data_dir'] = v
+            elif p == 'cluster_name' and v:
+                config['scope'] = v
+            elif p in ('archive_command', 'restore_command', 'archive_cleanup_command',
+                       'recovery_end_command', 'ssl_passphrase_command',
+                       'hba_file', 'ident_file', 'config_file'):
+                # write commands to the local config due to security implications
+                # write hba/ident/config_file to local config to ensure they are not removed later
+                config['postgresql'].setdefault('parameters', {})
+                config['postgresql']['parameters'][p] = v
+            elif p in helper_dict:
+                helper_dict[p] = v
+            else:
+                config['bootstrap']['dcs']['postgresql']['parameters'][p] = v
+
+    conn.close()
+
+    connect_port = parsed_dsn.get('port', os.getenv('PGPORT', helper_dict['port']))
+    config['postgresql']['connect_address'] = f'{get_ip()}:{connect_port}'
+    config['postgresql']['listen'] = f'{helper_dict["listen_addresses"]}:{helper_dict["port"]}'
+
+    # it makes sense to define postgresql.pg_hba/pg_ident only if hba_file/ident_file are set to defaults
```

Comment:
For sake of curiosity, do you know why Patroni ignores `postgresql.pg_hba` if `hba_file` in Postgres is not default?

Reply:
because it may happen it is located outside of PGDATA and we don't have write permissions for it

Reply:
Oh, I see. Thanks!